server/server.go

```diff
@@ -791,19 +781,30 @@ func (s *Server) UnixListenAndServe(path string) error {
 	return errors.New("can't listen on empty path")
 }
 
-// Close shuts down the listeners.
-func (s *Server) Close() {
+// Close shuts down the listeners and their active connections.
+func (s *Server) Close() error {
 	// Close each active listener. This will result in the blocking calls to
 	// Accept to immediately return with error, which will trigger the teardown
 	// of all active connections and associated goroutines.
 	s.mtx.Lock()
-	for _, l := range s.listeners {
+	defer s.mtx.Unlock()
+	if s.shutdown {
+		return fmt.Errorf("Close called multiple times")
+	}
+
+	s.shutdown = true
+	for l, conns := range s.listeners {
+		delete(s.listeners, l)
+
+		log.Debugf("Shutting down %v; closing %d active connections", l.Addr().String(), len(conns))
+		for conn := range conns {
+			conn.Destroy()
+		}
 		l.Close()
```

Comment:
Maybe we should close the listener socket ahead of sweeping through the connections. As is while the cleanup is happening lots of handshakes can still be happening that are sort of useless since they will be rejected immediately.

Reply:
I thought about that, but the advantage of doing it this way is it's clear in the logs why these connections are being closed. And in practice the way we do graceful reloads the new process should already be handling the new connections before we get here.

Reply:
Why would calling l.Close() before the loop change the logging? Not a big deal, I'm just a bit confused.

Reply:
B/c of LogConnErr:
```go
func (c *conn) LogConnErr(err error) {
	if err == nil { // We're destroying the connection
		log.Debugf("connection %v: server closing connection %s", c.name, c.stats)
	} else if err == io.EOF {
		log.Debugf("connection %v: closed by client %s", c.name, c.stats)
	} else {
		c.s.stats.logConnFailure()
		log.Errorf("connection %v: encountered error: %v %s", c.name, err, c.stats)
	}
}
```
It's the difference between logging "server closing connection" as a debug message and "encountered error" as an error message.

Reply:
I wrote a small program to test this and I'm wrong, closing the listener does not close the connections. Will fix.